File: c2-project/Solution55-2 2.cpp
//
// Created by 陈高杰 on 2020/2/17.
//
#include <iostream>
#include <stdlib.h>
#include <string>
#include <vector>
#include <stack>
#include <queue>
#include <math.h>
#include <sstream>
#include <map>
using namespace std;


struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};

class Solution {
public:
    bool IsBalanced_Solution(TreeNode* pRoot) {
        if(pRoot == nullptr)
            return true;
        if(TreeDepth(pRoot) !=-1)
            return true;
        else
            return false;
    }
    int TreeDepth(TreeNode* pRoot)
    {
        int depth = 0;
        if(pRoot == nullptr)
            return 0;
        int left = TreeDepth(pRoot->left);
        if(left ==-1)
            return -1;
        int right = TreeDepth(pRoot->right);
        if(right == -1)
            return -1 ;
        if(abs(left - right)>1)
            return -1;
        return (left > right ? left : right) + 1;
    }
};

//
//int main(){
//    TreeNode TreeNode1 = TreeNode(1);
//    TreeNode TreeNode2 = TreeNode(2);
//    TreeNode TreeNode3 = TreeNode(3);
//    TreeNode TreeNode4 = TreeNode(4);
//    TreeNode TreeNode5 = TreeNode(5);
//    TreeNode TreeNode6 = TreeNode(6);
//    TreeNode TreeNode7 = TreeNode(7);
//    TreeNode1.left = &TreeNode2;
//    TreeNode1.right = &TreeNode3;
//    TreeNode2.left = &TreeNode4;
//    TreeNode2.right = &TreeNode5;
//    TreeNode3.right = &TreeNode6;
//    TreeNode5.left = &TreeNode7;
//
//    Solution solution;
//    int k = 3;
//    int x = solution.IsBalanced_Solution(&TreeNode1);
//    cout<< x<<endl;
//    system("PAUSE");
//    return 0;
//}

File: py-project/Solution55-2 2.py
# -*- coding:utf-8 -*-
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def IsBalanced_Solution(self, pRoot):
        # write code here
        if pRoot == None:
            return True
        if self.dfs(pRoot)!=-1:
            return True
        else:
            return False

    def dfs(self,pRoot):
        if pRoot == None:
            return 0
        left = self.dfs(pRoot.left)
        if left == -1:
            return -1
        right = self.dfs(pRoot.right)
        if right == -1:
            return -1
        if abs(left-right)>1:
            return -1
        return max(left,right)+1


if __name__ == '__main__':
    TreeNode1 = TreeNode(5)
    TreeNode2 = TreeNode(3)
    TreeNode3 = TreeNode(7)
    TreeNode4 = TreeNode(2)
    TreeNode5 = TreeNode(4)
    TreeNode6 = TreeNode(6)
    TreeNode7 = TreeNode(8)
    TreeNode1.left = TreeNode2
    TreeNode1.right = TreeNode3
    TreeNode2.left = TreeNode4
    TreeNode2.right = TreeNode5
    TreeNode3.left = TreeNode6
    TreeNode3.right = TreeNode7
    k = 3
    # print(len(l))
    solution1 = Solution()
    a = solution1.IsBalanced_Solution(TreeNode1)
    print(a)